v2/internal/platform/systray/windows.go

```diff
@@ -0,0 +1,430 @@
+	}
+	return nil
+}
+
+func (p *Systray) newNotifyIconData() win32.NOTIFYICONDATA {
+	nid := win32.NOTIFYICONDATA{
+		UID:  p.id,
+		HWnd: p.hwnd,
+	}
+	nid.CbSize = uint32(unsafe.Sizeof(nid))
+	return nid
+}
+
+func (p *Systray) Show() error {
+	return p.setVisible(true)
+}
+
+func (p *Systray) Hide() error {
+	return p.setVisible(false)
+}
+
+func (p *Systray) setVisible(visible bool) error {
+	nid := p.newNotifyIconData()
+	nid.UFlags = win32.NIF_STATE
+	nid.DwStateMask = win32.NIS_HIDDEN
+	if !visible {
+		nid.DwState = win32.NIS_HIDDEN
+	}
+
+	if !win32.ShellNotifyIcon(win32.NIM_MODIFY, &nid) {
+		return errors.New("shell notify tooltip failed")
+	}
+	return nil
+}
+
+func (p *Systray) SetIcons(lightModeIcon, darkModeIcon *options.SystemTrayIcon) error {
+	var newLightModeIcon, newDarkModeIcon win32.HICON
+	if lightModeIcon != nil && lightModeIcon.Data != nil {
+		newLightModeIcon = p.getIcon(lightModeIcon.Data)
+	}
+	if darkModeIcon != nil && darkModeIcon.Data != nil {
+		newDarkModeIcon = p.getIcon(darkModeIcon.Data)
+	}
+	p.lightModeIcon, _ = lo.Coalesce(newLightModeIcon, newDarkModeIcon, p.appIcon)
+	p.darkModeIcon, _ = lo.Coalesce(newDarkModeIcon, newLightModeIcon, p.appIcon)
+	return p.updateIcon()
+}
+
+func (p *Systray) getIcon(icon []byte) win32.HICON {
+	result, err := win32.CreateHIconFromPNG(icon)
+	if err != nil {
+		result = p.appIcon
+	}
+	return result
+}
+
+func (p *Systray) setIcon(hicon win32.HICON) error {
+	nid := p.newNotifyIconData()
+	nid.UFlags = win32.NIF_ICON
+	if hicon == 0 {
+		nid.HIcon = 0
+	} else {
+		nid.HIcon = hicon
+	}
+
+	if !win32.ShellNotifyIcon(win32.NIM_MODIFY, &nid) {
+		return errors.New("shell notify icon failed")
+	}
+	return nil
+}
+
+func (p *Systray) WinProc(hwnd win32.HWND, msg uint32, wparam, lparam uintptr) uintptr {
+	switch msg {
+	case win32.NotifyIconMessageId:
+		switch lparam {
+		case win32.WM_LBUTTONUP:
+			if p.lclick != nil {
+				println("left click")
+				p.lclick()
+			}
+		case win32.WM_RBUTTONUP:
+			if p.rclick != nil {
+				println("right click")
+				p.rclick()
+			}
+		case win32.WM_LBUTTONDBLCLK:
+			if p.ldblclick != nil {
+				p.ldblclick()
+			}
+		case win32.WM_RBUTTONDBLCLK:
+			if p.rdblclick != nil {
+				p.rdblclick()
+			}
+		default:
+			//println(win32.WMMessageToString(lparam))
+		}
+	case win32.WM_SETTINGCHANGE:
+		settingChanged := win32.UTF16PtrToString(lparam)
+		if settingChanged == "ImmersiveColorSet" {
+			err := p.updateIcon()
+			if err != nil {
+				println("update icon failed", err.Error())
+			}
+		}
+		return 0
+	case win32.WM_COMMAND:
+		cmdMsgID := int(wparam & 0xffff)
+		switch cmdMsgID {
+		default:
+			p.menu.ProcessCommand(cmdMsgID)
+		}
+	default:
+		//msg := int(wparam & 0xffff)
+		//println(win32.WMMessageToString(uintptr(msg)))
+	}
+
+	result, _, _ := DefWindowProc.Call(uintptr(hwnd), uintptr(msg), wparam, lparam)
+	return result
+}
+
+func (p *Systray) Run() error {
+	var msg win32.MSG
```

Comment:
We might suffer here from the same problem as we had with our main message loop here: #1083

Reply:
Oooph that is a monumental catch 👍